Backpack/src/main/java/net/skyscanner/backpack/button/internal/BpkButtonBase.kt

```diff
@@ -0,0 +1,263 @@
+internal const val ICON_POSITION_ICON_ONLY = 2
+
+internal const val INVALID_RES = -1
+
+open abstract class BpkButtonBase internal constructor(
+  context: Context,
+  attrs: AttributeSet?,
+  defStyleAttr: Int
+) : AppCompatButton(wrapContextWithDefaults(context), attrs, defStyleAttr) {
+
+  private var isInitialized = false
+
+  private val textMeasurement: TextMeasurement by lazy { TextMeasurement(paint) }
+  internal val tokens = ButtonTokens(this.context)
+
+  private val font = BpkText.getFont(this.context, BpkText.SM, BpkText.Weight.EMPHASIZED)
+
+  private var overrideTextColor: Int? = null
+
+  private var _buttonTextColor: Int = ContextCompat.getColor(this.context, R.color.bpkWhite)
+  protected var buttonTextColor: Int
+    get() = overrideTextColor ?: _buttonTextColor
+    set(value) {
+      _buttonTextColor = value
+      updateSelf()
+    }
+
+  @Dimension
+  private var originalStartPadding: Int = 0
+
+  @Dimension
+  private var originalEndPadding: Int = 0
+
+  private var _iconPosition: Int = ICON_POSITION_END
+  open var iconPosition: Int
+    get() = _iconPosition
+    set(value) {
+      _iconPosition = value
+      updateSelf()
+    }
+
+  private var _icon: Drawable? = null
+  var icon: Drawable? = null
+    get() = _icon
+    set(value) {
+      if (value != field) {
+        field = value
+
+        value?.let {
+          _icon = convertAndResize(it)
+        }
+
+        updateSelf()
+      }
+    }
+
+  init {
+    maxLines = 1
+    isAllCaps = true
+    gravity = Gravity.CENTER
+
+    val attr = this.context.theme.obtainStyledAttributes(attrs, net.skyscanner.backpack.R.styleable.BpkButton, defStyleAttr, 0)
+    attr.getColor(net.skyscanner.backpack.R.styleable.BpkButton_buttonTextColor, INVALID_RES).let {
+      if (it != INVALID_RES) {
+        overrideTextColor = it
+      }
+    }
+
+    _iconPosition = attr.getInt(R.styleable.BpkButton_buttonIconPosition, ICON_POSITION_END)
+    attr.getResourceId(R.styleable.BpkButton_buttonIcon, INVALID_RES).let {
+      if (it != INVALID_RES) {
+        _icon = AppCompatResources.getDrawable(context, it)?.let(::convertAndResize)
+      }
+    }
+
+    attr.recycle()
+    updateSelf()
+
+    isInitialized = true
+  }
+
+  private fun updateSelf(triggerUpdate: Boolean = false) {
+    isClickable = isEnabled
+
+    font.applyTo(this)
+
+    if (this.isEnabled) {
+      this.setTextColor(
+        getColorSelector(
+          buttonTextColor,
+          darken(buttonTextColor, .1f),
+          tokens.gray300
+        )
+      )
+    } else {
+      this.setTextColor(tokens.gray300)
+    }
+
+    _icon?.let {
+      DrawableCompat.setTintList(
+        it,
+        getColorSelector(
+          buttonTextColor,
+          darken(buttonTextColor, .1f),
+          tokens.gray300
+        )
+      )
+
+      this.setCompoundDrawablesRelativeWithIntrinsicBounds(
+        if (iconPosition == ICON_POSITION_START || iconPosition == ICON_POSITION_ICON_ONLY) it else null,
```

Comment:
it.takeIf { iconPosition == ICON_POSITION_START || iconPosition == ICON_POSITION_ICON_ONLY }

Reply:
done